src/volume/inmem.rs

```diff
@@ -103,9 +108,17 @@ impl InMemNiftiVolume {
     /// of the volume's data must be known in advance. It it also expected that the
     /// following bytes represent the first voxels of the volume (and not part of the
     /// extensions).
-    pub fn from_reader<R: Read>(mut source: R, header: &NiftiHeader) -> Result<Self> {
-        let mut raw_data = vec![0u8; nb_bytes_for_data(header)?];
-        source.read_exact(&mut raw_data)?;
+    pub fn from_reader<R: Read>(source: R, header: &NiftiHeader) -> Result<Self> {
+        // rather than pre-allocating for the full volume size, this will
+        // pre-allocate up to a more reliable amount and feed the vector
+        // sequentially, to prevent some trivial OOM attacks
+        let nb_bytes = nb_bytes_for_data(header)?;
+        let mut raw_data = Vec::with_capacity(nb_bytes.min(PREALLOC_MAX_SIZE));
+        let nb_bytes_written = std::io::copy(&mut source.take(nb_bytes as u64), &mut raw_data)?;
+
+        if nb_bytes_written as usize != nb_bytes {
```

Comment:
Too much casts here. A single

    let nb_bytes = nb_bytes_for_data(header)? as u64;

intead of `nb_bytes as u64` and `nb_bytes_written as usize`. Same for above.

Reply:
I could only deduplicate the cast of `nb_bytes_written` to `usize` once. If I turn `nb_bytes` to `u64`, it will have to be converted back to a `usize` at `Vec::with_capacity` and `NiftiError::IncompatibleLength`.